KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,714 @@
+     */
+    function _validateAdmin(
+        address _approver,
+        address[] memory adminList
+    ) private returns (bool isAdmin) {
+        for (uint256 i = 0; i < adminList.length; i++) {
+            if (_approver == adminList[i]) {
+                isAdmin = true;
+            }
+        }
+    }
+
+    /**
+     * @dev gets the adminList and quorom by calling `getState()` method in retiredAddress contract
+     * @param _retiredAddress is the address of the contract
+     * @return adminList list of the retiredAddress contract admins
+     * @return req min required number of approvals
+     */
+    function _getState(
+        address _retiredAddress
+    ) private returns (address[] memory adminList, uint256 req) {
+        //call getState() function in retiredAddress contract to get the adminList
+        bytes memory payload = abi.encodeWithSignature("getState()");
+        (bool success, bytes memory result) = _retiredAddress.staticcall(
+            payload
+        );
+        emit RetiredAddrAdminStatus(success, result);
+        require(success, "call failed");
+
+        (adminList, req) = abi.decode(result, (address[], uint256));
+    }
+
+    /**
+     * @dev Internal function to update the approver details of a retired
+     * _retiredAddress is the address of the retired
+     * _approver is the admin of the retiredAddress
+     */
+    function _updateApprover(
+        address _retiredAddress,
+        address _approver
+    ) private {
+        uint256 index = getRetiredIndex(_retiredAddress);
+        address[] memory approvers = retirees[index].approvers;
+        for (uint256 i = 0; i < approvers.length; i++) {
+            require(approvers[i] != _approver, "Already approved");
+        }
+        retirees[index].approvers.push(_approver);
+        emit Approved(
+            _retiredAddress,
+            _approver,
+            retirees[index].approvers.length
+        );
+    }
+
+    /**
+     * @dev finalizeRegistration sets the status to Registered,
+     *      After this stage, registrations will be restricted.
+     */
+    function finalizeRegistration()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Initialized)
+    {
+        status = Status.Registered;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev finalizeApproval sets the status to Approved,
+     *      After this stage, approvals will be restricted.
+     */
+    function finalizeApproval()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Registered)
+    {
+        require(
+            getTreasuryAmount() < sumOfRetiredBalance(),
+            "treasury amount should be less than the sum of all retired address balances"
+        );
+        _isRetiredsApproved();
+        status = Status.Approved;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev verify if quorom reached for the retired approvals
+     */
+    function _isRetiredsApproved() private {
```

Comment:
Can we make this function as a view function since it does not update any storage..?

Reply:
changed